iothub/device/devdoc/API Design.md

```diff
@@ -6,297 +6,142 @@
 }
-```
-</details>
-
-
-<details>
-<summary>WritableProperty</summary>
-
-```csharp
-/// <summary>
-/// Empty constructor.
-/// </summary>
-public WritablePropertyResponse() { }
 
-/// <summary>
-/// Convenience constructor for specifying the properties.
-/// </summary>
-/// <param name="propertyValue">The unserialized property value.</param>
-/// <param name="ackCode">The acknowledgement code, usually an HTTP Status Code e.g. 200, 400.</param>
-/// <param name="ackVersion">The acknowledgement version, as supplied in the property update request.</param>
-/// <param name="ackDescription">The acknowledgement description, an optional, human-readable message about the result of the property update.</param>
-public WritablePropertyResponse(object propertyValue, int ackCode, long ackVersion, string ackDescription = null)
-{
-    PropertyValue = propertyValue;
-    AckCode = ackCode;
-    AckVersion = ackVersion;
-    AckDescription = ackDescription;
+public class WritablePropertyResponse {
+    public WritablePropertyResponse(object propertyValue, PropertyConvention propertyConvention=null);
+    public WritablePropertyResponse(object propertyValue, int ackCode, long ackVersion, string ackDescription=null, PropertyConvention propertyConvention=null);
+    public int AckCode { get; set; }
+    public string AckDescription { get; set; }
+    public long AckVersion { get; set; }
+    public object Value { get; private set; }
+    public JRaw ValueAsJson { get; }
 }
 
-/// <summary>
-/// The unserialized property value.
-/// </summary>
-[JsonProperty("value")]
-public object PropertyValue { get; set; }
-
-/// <summary>
-/// The acknowledgement code, usually an HTTP Status Code e.g. 200, 400.
-/// </summary>
-[JsonProperty("ac")]
-public int AckCode { get; set; }
-
-/// <summary>
-/// The acknowledgement version, as supplied in the property update request.
-/// </summary>
-[JsonProperty("av")]
-public long AckVersion { get; set; }
+public class ObjectSerializer {
+    public static readonly ObjectSerializer Instance;
+    public ObjectSerializer();
+    public string ContentType { get; set; }
+    public virtual T DeserializeToType<T>(string stringToDeserialize);
+    public virtual string SerializeToString(object objectToSerialize);
+}
 
-/// <summary>
-/// The acknowledgement description, an optional, human-readable message about the result of the property update.
-/// </summary>
-[JsonProperty("ad", DefaultValueHandling = DefaultValueHandling.Ignore)]
-public string AckDescription { get; set; }
+public class PropertyConvention {
+    public static readonly PropertyConvention Instance;
+    public PropertyConvention();
+    public Encoding ContentEncoding { get; }
+    public string ContentType { get; }
+    public ObjectSerializer PayloadSerializer { get; set; }
+    public static PropertyCollection CreatePropertiesPatch(IDictionary<string, object> properties, string componentName=null, PropertyConvention propertyConvention=null);
```

Comment:
API takes in IConventionHandler which is composed of ISerializer and IContentEncoder